### PR TITLE
Add Time class and tests

### PR DESCRIPTION
**Pull request by jzwang8**
Time class used in FindMatchQuery tests.

### CODE REVIEW

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    if (minutes < 0 || minutes >= 60) {
+      throw new IllegalArgumentException("Minutes can only be 0 through 59 (inclusive).");
+    }
+
+    // Calculate today's date but with hours:minutes time
+    // TODO: All times are currently today, wrap around times?
+    Calendar c = Calendar.getInstance();
+		c.setTime(date);
```
Comment (dhruvi618) [inline backend/src/main/java/com/google/sps/data/Time.java:35]: nit: indentation

File: backend/src/test/java/com/google/sps/TimeTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    Calendar c = Calendar.getInstance();
+    c.set(2020, 0, 1, 14, 0, 0);
+    c.set(Calendar.MILLISECOND, 0);
+    Date date = c.getTime();
+
+    time = new Time(date);
+  }
+
+  @Test
+  public void inputTimeToMillis() {
```
Comment (dhruvi618) [inline backend/src/test/java/com/google/sps/TimeTest.java:29]: Since the getTimeMillis() method has two edge cases that check for out of bound hours/minutes, would it be possible to add tests that ensure an IllegalArgumentException is thrown?

File: backend/src/test/java/com/google/sps/TimeTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/** */
+public final class TimeTest {
+
+  private Time time;
+
+  @Before
+  public void setUp() {
+    // Set "current" date to  1/1/2020 2:00pm GMT
+    Calendar c = Calendar.getInstance();
+    c.set(2020, 0, 1, 14, 0, 0);
```
Comment (dhruvi618) [inline backend/src/test/java/com/google/sps/TimeTest.java:21]: What do you think about using a date formatter to try to establish a common way to set dates and potentially catch any date errors? This is the documentation for the [SimpleDateFormat](https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html) and I also found an [example](https://mkyong.com/java/java-convert-date-and-time-between-timezone/) of how it's used with the Calendar and other classes since I know in the future we might need to convert between timezones.

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  public Time() {
+    date = new Date();
+  }
+
+  /** Constructor with manually set date */
+  public Time(Date date) {
+    this.date = date;
+  }
+
+  /** Return today's date with time of hours:minutes in milliseconds since epoch */
```
Comment (dhruvi618) [inline backend/src/main/java/com/google/sps/data/Time.java:21]: nit: I think the `/**` makes it a javadoc instead of a comment

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package com.google.sps.data;
+
+import java.util.Date;
+import java.util.Calendar;
+
+/** Class to convert Time */
+public final class Time {
+
+  private static Date date;
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/data/Time.java:9]: why static?

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    // Calculate today's date but with hours:minutes time
+    // TODO: All times are currently today, wrap around times?
+    Calendar c = Calendar.getInstance();
+		c.setTime(date);
+    c.set(Calendar.HOUR_OF_DAY, hours);
+    c.set(Calendar.MINUTE, minutes);
+    c.set(Calendar.SECOND, 0);
+    c.set(Calendar.MILLISECOND, 0);
+    return c.getTime().getTime();
```
Comment (tcwanggoogle) [inline backend/src/main/java/com/google/sps/data/Time.java:40]: (Just want to make sure it's not a typo) Do we really need to `getTime()` twice?

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package com.google.sps.data;
+
+import java.util.Date;
+import java.util.Calendar;
+
+/** Class to convert Time */
+public final class Time {
```
Comment (tcwanggoogle) [inline backend/src/main/java/com/google/sps/data/Time.java:7]: (A question from someone who isn't familiar with Java) This class seems to implement some very common functionalities. Don't we have standard Java libraries to do similar things, or can we try to use standard library to avoid writing new functions if possible?

File: backend/src/test/java/com/google/sps/TimeTest.java (hunk 2 above)
Comment (tcwanggoogle) [reply backend/src/test/java/com/google/sps/TimeTest.java:29]: +1 to add tests covering corner cases.

File: backend/src/test/java/com/google/sps/TimeTest.java (hunk 3 above)
Comment (tcwanggoogle) [reply backend/src/test/java/com/google/sps/TimeTest.java:21]: +1 to encourage using standard libraries whenever possible

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 7 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java:7]: The main method (getTimeInMillis()) is pretty specific to our implementation. It takes an input "time" such as 11:00am and converts that to a Date timestamp on the current date but at that time. This implementation isn't final, though, because it all depends on what input we get from the front end input form.

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 6 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java:39]: The first getTime() of Calendar returns a Date object, and the second getTime() of Date returns a long timestamp.

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 5 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java]: I got rid of this static, made the getTimeMillis() method static, and changed the references in TimeTest.java

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+  private static Date date;
+
+  /** Constructor */
+  public Time() {
+    date = new Date();
+  }
```
[the remark this answers is not in the document]
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java]: Changed it so that this constructor called the other. Is it correct now?

File: backend/src/main/java/com/google/sps/data/Time.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
+package com.google.sps.data;
+
+import java.util.Date;
+import java.util.Calendar;
+
+/** Class to convert Time */
```
[the remark this answers is not in the document]
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java]: Explained. Let me know if it's still unclear. Basically, it's a class I'm using for both testing (setting a new reference date) and converting user input to an actual date.

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 4 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java:21]: I meant to have it be a javadoc. Should it not be?

File: backend/src/test/java/com/google/sps/TimeTest.java (hunk 3 above)
Comment (jzwang8) [reply backend/src/test/java/com/google/sps/TimeTest.java]: Right now for the prototype, everything is in the long timestamp format, but for the MVP and depending on user input format, I'll make it a more standardized time format.

File: backend/src/test/java/com/google/sps/TimeTest.java (hunk 2 above)
Comment (jzwang8) [reply backend/src/test/java/com/google/sps/TimeTest.java:29]: Added tests, thanks for the suggestion!

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 1 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java]: For some reason, that weird indentation doesn't show in my editor?

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 6 above)
Comment (tcwanggoogle) [reply backend/src/main/java/com/google/sps/data/Time.java:39]: Ok. Interesting. I wonder why the first is not called `getDate`, which would make more sense. But that's not your fault

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 1 above)
Comment (tcwanggoogle) [reply backend/src/main/java/com/google/sps/data/Time.java]: Hopefully with formatter in place, we won't see these problems in the future.

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 7 above)
Comment (tcwanggoogle) [reply backend/src/main/java/com/google/sps/data/Time.java:7]: Cool. So maybe explain a bit about this special date + time combination in the javadoc? (The name "Time" makes me think it's kind of a standard interpretation of Time)

Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java:7]: Changed the method name to getDateMillis() to clarify.

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 5 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/Time.java:9]: Edit: I changed it back to being static for it to be referenced easily in the Test. For now, this method is only used in the tests, but I also wanted it to be used when setting timeAvailableUntil later.

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 7 above)
Comment (kmh287) [reply backend/src/main/java/com/google/sps/data/Time.java:7]: I agree with Tiancong that "Time" might be too generic a name for this class. 

I'm wondering though if we could leverage the java 8 time libraries to simplify this somewhat. Could we implement similar functionality with:

```
long timestamp = LocalDate.now().atTime(hour, minute).toEpochSecond(ZoneOffset.UTC)
```
https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html

File: backend/src/test/java/com/google/sps/TimeTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    c.set(Calendar.MINUTE, 30);
+    Assert.assertEquals(c.getTime().getTime(), Time.getDateMillis(10, 30));
+
+    // 1/1/2020 11:45pm GMT
+    c.set(Calendar.HOUR_OF_DAY, 23);
+    c.set(Calendar.MINUTE, 45);
+    Assert.assertEquals(c.getTime().getTime(), Time.getDateMillis(23, 45));
+  }
+
+  @Test (expected = IllegalArgumentException.class)
```
Comment (kmh287) [inline backend/src/test/java/com/google/sps/TimeTest.java:52]: nice!

File: backend/src/test/java/com/google/sps/TimeTest.java (hunk 3 above)
Comment (dhruvi618) [reply backend/src/test/java/com/google/sps/TimeTest.java]: Ok, sounds good

File: backend/src/main/java/com/google/sps/data/Time.java (hunk 4 above)
Comment (dhruvi618) [reply backend/src/main/java/com/google/sps/data/Time.java:21]: Nevermind, it should be. Sorry for the confusion.
